diff -Nru xfprint-4.2.1.orig/xfprint/Makefile.in xfprint-4.2.1/xfprint/Makefile.in
--- xfprint-4.2.1.orig/xfprint/Makefile.in	2005-04-17 11:33:16.000000000 +0200
+++ xfprint-4.2.1/xfprint/Makefile.in	2005-04-17 12:33:43.000000000 +0200
@@ -147,7 +147,7 @@
 INSTALL_STRIP_PROGRAM = @INSTALL_STRIP_PROGRAM@
 INSTOBJEXT = @INSTOBJEXT@
 INTLLIBS = @INTLLIBS@
-LDFLAGS = @LDFLAGS@
+LDFLAGS = @LDFLAGS@ -Wl,-E
 LIBOBJS = @LIBOBJS@
 LIBS = @LIBS@
 LIBTOOL = @LIBTOOL@
diff -Nru xfprint-4.2.1.orig/xfprint-manager/Makefile.in xfprint-4.2.1/xfprint-manager/Makefile.in
--- xfprint-4.2.1.orig/xfprint-manager/Makefile.in	2005-04-17 11:33:16.000000000 +0200
+++ xfprint-4.2.1/xfprint-manager/Makefile.in	2005-04-17 12:34:10.000000000 +0200
@@ -146,7 +146,7 @@
 INSTALL_STRIP_PROGRAM = @INSTALL_STRIP_PROGRAM@
 INSTOBJEXT = @INSTOBJEXT@
 INTLLIBS = @INTLLIBS@
-LDFLAGS = @LDFLAGS@
+LDFLAGS = @LDFLAGS@ -Wl,-E
 LIBOBJS = @LIBOBJS@
 LIBS = @LIBS@
 LIBTOOL = @LIBTOOL@
