diff -ru bzip2-1.0.2/Makefile bzip2-1.0.2-new/Makefile
--- bzip2-1.0.2/Makefile	Sat Jan 26 00:34:53 2002
+++ bzip2-1.0.2-new/Makefile	Sun Feb 10 00:32:38 2002
@@ -9,7 +9,7 @@
 
 # Suitably paranoid flags to avoid bugs in gcc-2.7
 BIGFILES=-D_FILE_OFFSET_BITS=64
-CFLAGS=-Wall -Winline -O2 -fomit-frame-pointer -fno-strength-reduce $(BIGFILES)
+CFLAGS+=-Wall -Winline -fomit-frame-pointer -fno-strength-reduce $(BIGFILES)
 
 # Where you want it installed when you do 'make install'
 PREFIX=/usr
