diff -Nru autofs-4.1.4-old/samples/rc.autofs.in autofs-4.1.4/samples/rc.autofs.in
--- autofs-4.1.4-old/samples/rc.autofs.in	2005-05-15 13:30:40.000000000 +0200
+++ autofs-4.1.4/samples/rc.autofs.in	2005-05-15 13:51:07.000000000 +0200
@@ -43,6 +43,8 @@
 	system=debian
 elif [ -f /etc/redhat-release ]; then
 	system=redhat
+elif [ -f /usr/bin/crux ]; then
+	system=crux
 else
 	echo "$0: Unknown system, please port and contact autofs@linux.kernel.org" 1>&2
 	exit 1
@@ -104,6 +106,25 @@
 	[ -f /etc/default/autofs ] && . /etc/default/autofs
 
 	daemonoptions="$daemonoptions --timeout=$TIMEOUT"
+elif [ "$system" = "crux" ]; then
+        LOCALOPTIONS=""
+        DAEMONOPTIONS=""
+        UNDERSCORETODOT=1
+        DISABLE_DIRECT=1
+        DAEMON_EXIT_WAIT=20
+	TIMEOUT=300
+
+	[ -f /etc/autofs/config ] && . /etc/autofs/config
+
+	# Over-ride localoptions if set
+	if [ -n "$LOCALOPTIONS" ]; then
+		localoptions=$LOCALOPTIONS
+	fi
+
+	# Over-ride daemonoptions if set
+	if [ -n "$DAEMONOPTIONS" ]; then
+		daemonoptions=$DAEMONOPTIONS
+	fi
 fi
 
 #
@@ -433,7 +454,7 @@
 		done
 	done
 
-	if [ -n "`/sbin/pidof $DAEMON`" ] ; then
+	if [ -n "`/bin/pidof $DAEMON`" ] ; then
 		RETVAL=1
 	fi
 
@@ -530,15 +551,15 @@
 		RETVAL=$?
 	fi
 	count=0
-	while [ -n "`/sbin/pidof $DAEMON`" -a $count -lt 10 ] ; do
+	while [ -n "`/bin/pidof $DAEMON`" -a $count -lt 10 ] ; do
 	   killproc $DAEMON -USR2 >& /dev/null
 	   RETVAL=$?
-	   [ $RETVAL = 0 -a -z "`/sbin/pidof $DAEMON`" ] || sleep 3
+	   [ $RETVAL = 0 -a -z "`/bin/pidof $DAEMON`" ] || sleep 3
 	   count=`expr $count + 1`
 	done
 	umount -a -f -t autofs
 	rm -f /var/lock/subsys/autofs
-	if [ -n "`/sbin/pidof $DAEMON`" ] ; then
+	if [ -n "`/bin/pidof $DAEMON`" ] ; then
 	    failure "$prog shutdown"
 	else
 	    success "$prog shutdown"
@@ -766,11 +787,62 @@
 esac
 }
 
+function crux()
+{
+case "$1" in
+
+start)
+	# Make sure the autofs filesystem type is available.
+	(grep -q autofs /proc/filesystems || /sbin/modprobe -k autofs4 || /sbin/modprobe -k autofs) 2> /dev/null
+	TMP=`mktemp /tmp/autofs.XXXXXX` || { echo $"could not make temp file" >& 2; exit 1; }
+	getmounts | tee $TMP | sh
+	RETVAL=$?
+	if [ ! -s $TMP ] ; then
+		echo "No Mountpoints Defined" >& 2
+	fi
+	rm -f $TMP
+	;;
+
+stop)
+	signal_automounts
+	RETVAL=$?
+	count=0
+	while [ -n "`/bin/pidof $DAEMON`" -a $count -lt 5 ] ; do
+		killall -USR2 $DAEMON >& /dev/null
+		RETVAL=$?
+		[ $RETVAL = 0 -a -z "`/bin/pidof $DAEMON`" ] || sleep .3 
+		count=`expr $count + 1`
+	done
+	umount -a -f -t autofs 2> /dev/null 
+	RETVAL=$?
+	;;
+getmounts)
+        getmounts
+        ;;
+status)
+	status
+	;;
+restart)
+	crux stop
+	crux start
+	;;
+active)
+	alive
+	;;
+*)
+	echo $"Usage: $0 {start|stop|restart|status|getmounts|active}"
+	RETVAL=0
+esac
+
+}
+
 RETVAL=0
 if [ $system = debian ]; then
 	debian "$@"
 elif [ $system = redhat ]; then
 	redhat "$@"
+elif [ $system = crux ]; then
+	crux "$@"
 fi
 
 exit $RETVAL
