--- Makefile.orig	2005-04-15 12:53:20.400368552 -0500
+++ Makefile	2005-04-15 12:53:30.975760848 -0500
@@ -250,8 +250,8 @@
 	@echo \#define UDEV_CONFIG_DIR		\"$(configdir)\" >> $@
 	@echo \#define UDEV_CONFIG_FILE		\"$(configdir)/udev.conf\" >> $@
 	@echo \#define UDEV_RULES_FILE		\"$(configdir)/rules.d\" >> $@
-	@echo \#define UDEV_BIN			\"$(DESTDIR)$(sbindir)/udev\" >> $@
-	@echo \#define UDEVD_BIN		\"$(DESTDIR)$(sbindir)/udevd\" >> $@
+	@echo \#define UDEV_BIN			\"$(sbindir)/udev\" >> $@
+	@echo \#define UDEVD_BIN		\"$(sbindir)/udevd\" >> $@
 
 # Rules on how to create the generated config files
 $(LOCAL_CFG_DIR)/udev.conf:
