diff -Nru wvdial-1.54.0.orig/Makefile wvdial-1.54.0/Makefile
--- wvdial-1.54.0.orig/Makefile	2005-05-09 19:21:39.000000000 +0200
+++ wvdial-1.54.0/Makefile	2005-05-09 19:52:18.000000000 +0200
@@ -1,15 +1,17 @@
 ifeq ($(TOPDIR),)
   TOPDIR=.
-  PKGINC=/usr/include/wvstreams /usr/local/include/wvstreams
+  PKGINC=$(WVS)/include
   LIBS := $(LIBS) \
 		$(shell $(CC) -lsupc++ 2>&1 | grep -q "undefined reference" \
 			&& echo " -lsupc++")
 endif
 
-PREFIX=/usr/local
+PREFIX=$(DESTDIR)/usr
 BINDIR=${PREFIX}/bin
-MANDIR=${PREFIX}/share/man
-PPPDIR=/etc/ppp/peers
+MANDIR=${PREFIX}/man
+PPPDIR=$(DESTDIR)/etc/ppp/peers
+
+VERBOSE=1
 
 include $(TOPDIR)/wvrules.mk
 
@@ -21,7 +23,7 @@
 wvdial.a: wvdialer.o wvdialtext.o wvmodemscan.o wvpapchap.o wvdialbrain.o \
 	wvdialmon.o
 
-LIBS += -L../wvstreams -lwvutils -lwvstreams
+LIBS += $(WVS)/libwvstreams.a $(WVS)/libwvutils.a $(WVS)/libwvbase.a $(WVS)/xplc/libxplc-cxx.a
 
 wvdial wvdialconf papchaptest pppmon: wvdial.a
 
